File: packages/eyes-universal/src/socket.js
const WebSocket = require('ws')
const uuid = require('uuid')

function makeSocket(ws) {
  let socket = null
  const listeners = new Map()
  const queue = new Set()

  attach(ws)

  function connect(url) {
    const ws = new WebSocket(url)
    attach(ws)
  }

  function attach(ws) {
    if (!ws) return

    if (ws.readyState === WebSocket.CONNECTING) ws.on('open', () => attach(ws))
    else if (ws.readyState === WebSocket.OPEN) {
      socket = ws
      queue.forEach(command => command())
      queue.clear()

      socket.on('message', message => {
        const {name, requestId, payload} = deserialize(message)
        const fns = listeners.get(name)
        if (fns) fns.forEach(fn => fn(payload, requestId))
      })
      socket.on('close', () => {
        const fns = listeners.get('close')
        if (fns) fns.forEach(fn => fn())
      })
    }
  }

  function disconnect() {
    if (!socket) return
    socket.terminate()
    socket = null
  }

  function request(name, payload) {
    return new Promise((resolve, reject) => {
      const requestId = uuid.v4()
      emit({name, requestId}, payload)
      once({name, requestId}, response => {
        if (response.error) return reject(response.error)
        return resolve(response.result)
      })
    })
  }

  function command(name, fn) {
    on(name, async (payload, requestId) => {
      try {
        const result = await fn(payload)
        emit({name, requestId}, {result})
      } catch (error) {
        emit({name, requestId}, {error})
      }
    })
  }

  function on(type, fn) {
    type = typeof type === 'string' ? {name: type} : {name: type.name, requestId: type.requestId}
    let fns = listeners.get(type.name)
    if (!fns) {
      fns = new Set()
      listeners.set(type.name, fns)
    }
    const handler = type.requestId
      ? (payload, requestId) => requestId === type.requestId && fn(payload, requestId)
      : fn
    fns.add(handler)
    return () => off(type.name, handler)
  }

  function once(type, fn) {
    const off = on(type, (...args) => (fn(...args), off()))
    return off
  }

  function off(name, fn) {
    if (!fn) return listeners.delete(name)
    const fns = listeners.get(name)
    if (!fns) return false
    const existed = fns.delete(fn)
    if (!fns.size) listeners.delete(name)
    return existed
  }

  function emit(type, payload) {
    const command = () => socket.send(serialize(type, payload))
    if (socket) command()
    else queue.add(command)
    return () => queue.delete(command)
  }

  function ref() {
    const command = () => socket._socket.ref()
    if (socket) command()
    else queue.add(command)
    return () => queue.delete(command)
  }

  function unref() {
    const command = () => socket._socket.unref()
    if (socket) command()
    else queue.add(command)
    return () => queue.delete(command)
  }

  return {
    on,
    once,
    off,
    emit,
    request,
    command,
    connect,
    disconnect,
    ref,
    unref,
  }
}

function serialize(type, payload) {
  const message =
    typeof type === 'string'
      ? {name: type, payload}
      : {name: type.name, requestId: type.requestId, payload}
  return JSON.stringify(message)
}

function deserialize(message) {
  return JSON.parse(message)
}

module.exports = makeSocket

File: rb/eyes-selenium/lib/applitools/socket.rb
require('faye/websocket')
require('json')
require('securerandom')
require('colorize')
require_relative('./logger')

module Applitools
  class Socket
    attr_reader :listeners, :queue
    include ::Applitools::Logger

    def initialize
      @listeners = {}
      @queue = []
    end

    def connect(uri, ws = ::Faye::WebSocket::Client.new(uri))
      @socket = ws

      queue.each {|command| command.call}
      queue.clear

      ws.on :message do |event|
        message = JSON.parse(event.data, {:symbolize_names => true})
        params = [message[:payload], message[:key]]
        find_and_execute_listeners(message[:name], message[:key], params)
      end

      ws.on :close do |event|
        find_and_execute_listeners('close')
      end
    end

    def emit(message, payload)
      command = ->() {@socket.send(serialize(message, payload))}
      @socket ? command.call : queue.push(command)
    end

    def command(name, fn)
      on(name, ->(payload, key) {
        begin
          log("[#{'COMMAND'.yellow}] #{name}, #{key}, #{JSON.pretty_generate(payload)}")
          result = fn.call(payload)
          emit({name: name, key: key}, {result: result})
        rescue => error
          log("[#{'COMMAND ERROR'.red}] #{error}")
          emit({name: name, key: key}, error.message || error)
        end
      })
    end

    def request(name, payload, key = SecureRandom.uuid, cb = nil)
      log("[#{'REQUEST'.blue}] #{name}, #{key}, #{JSON.pretty_generate(payload)}")
      emit({name: name, key: key}, payload)
      once({name: name, key: key}, Proc.new {|result|
        cb.call(result[:result] || result[:error] || true) if cb
      })
    end

    private

      def find_and_execute_listeners(name, key = nil, params = [])
        name_with_key = "#{name}/#{key}"
        fns = listeners[name.to_sym]
        fns = listeners[name_with_key.to_sym] if (!fns) 
        return if (!fns)
        fns.each {|fn| fn.call(*params)}
      end

      def serialize(type, payload)
        message = type.is_a?(String) ? 
          {:name => type, :payload => payload} : {:name => type[:name], key: type[:key], :payload => payload}
        JSON.generate(message)
      end

      def get_name_from_type(type)
        type.is_a?(String) ? type : "#{type[:name]}/#{type[:key]}"
      end

      def on(type, fn)
        name = get_name_from_type(type)
        fns = listeners[name.to_sym]
        if (!fns)
          fns = []
          listeners[name.to_sym] = fns
        end
        fns.push(fn)
      end

      def off(type)
        name = get_name_from_type(type)
        listeners.delete(name.to_sym)
      end

      def once(type, fn)
        on(type, ->(*args) {
          fn.call(*args)
          off(type)
        })
      end
  end
end
